From 8866b11a67797e3812e9685c4269ddd54f8d4948 Mon Sep 17 00:00:00 2001
From: Akira Ajisaka <aajisaka@apache.org>
Date: Tue, 26 May 2015 16:16:21 +0900
Subject: [PATCH 1451/2748] HADOOP-8751. NPE in Token.toString() when Token is
 constructed using null identifier. Contributed by
 kanaka kumar avvaru.

(cherry picked from commit 56996a685e6201cb186cea866d22418289174574)
(cherry picked from commit 70615947bdac9c4050a00465c426de9e783d0d74)

Change-Id: I4ea5d32bd1f55218528e534d7d5dd48952c78916
---
 .../org/apache/hadoop/security/token/Token.java    |    8 ++++----
 .../token/delegation/TestDelegationToken.java      |   15 +++++++++++++++
 2 files changed, 19 insertions(+), 4 deletions(-)

diff --git a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/token/Token.java b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/token/Token.java
index 6403b22..d36f96f 100644
--- a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/token/Token.java
+++ b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/token/Token.java
@@ -73,10 +73,10 @@ public Token(T id, SecretManager<T> mgr) {
    * @param service the service for this token
    */
   public Token(byte[] identifier, byte[] password, Text kind, Text service) {
-    this.identifier = identifier;
-    this.password = password;
-    this.kind = kind;
-    this.service = service;
+    this.identifier = (identifier == null)? new byte[0] : identifier;
+    this.password = (password == null)? new byte[0] : password;
+    this.kind = (kind == null)? new Text() : kind;
+    this.service = (service == null)? new Text() : service;
   }
 
   /**
diff --git a/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/security/token/delegation/TestDelegationToken.java b/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/security/token/delegation/TestDelegationToken.java
index 239b841..b41ff15 100644
--- a/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/security/token/delegation/TestDelegationToken.java
+++ b/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/security/token/delegation/TestDelegationToken.java
@@ -44,6 +44,7 @@
 import org.apache.hadoop.security.token.SecretManager;
 import org.apache.hadoop.security.token.Token;
 import org.apache.hadoop.security.token.SecretManager.InvalidToken;
+import org.apache.hadoop.security.token.TokenIdentifier;
 import org.apache.hadoop.security.token.delegation.AbstractDelegationTokenSecretManager.DelegationTokenInformation;
 import org.apache.hadoop.util.Daemon;
 import org.apache.hadoop.util.Time;
@@ -539,4 +540,18 @@ public void testDelegationKeyEqualAndHash() {
     Assert.assertEquals(key1, key2);
     Assert.assertFalse(key2.equals(key3));
   }
+
+  @Test
+  public void testEmptyToken() throws IOException {
+    Token<?> token1 = new Token<TokenIdentifier>();
+
+    Token<?> token2 = new Token<TokenIdentifier>(new byte[0], new byte[0],
+        new Text(), new Text());
+    assertEquals(token1, token2);
+    assertEquals(token1.encodeToUrlString(), token2.encodeToUrlString());
+
+    token2 = new Token<TokenIdentifier>(null, null, null, null);
+    assertEquals(token1, token2);
+    assertEquals(token1.encodeToUrlString(), token2.encodeToUrlString());
+  }
 }
-- 
1.7.9.5

